controller/attribute/util.py

```diff
@@ -71,12 +80,222 @@ def prepare_sample_records_doc_bin(attribute_id: str, project_id: str) -> str:
+        raise LlmResponseError(
+            "User prompt does not carry a single valid Mustache syntax for attribute access. "
+            "You can access attributes by using '{{ attribute_name }}' in your prompt."
+        )
+
+    for attr in mustache_attributes:
+        if not attribute.get_by_name(project_id, attr):
+            raise LlmResponseError(f"Attribute '{attr}' does not exist in the project.")
+
+
+def validate_llm_config(llm_config: Dict[str, Any]):
+    # test LLM connection before sending work package to execution environment
+    try:
+        if llm_config["llmIdentifier"] == enums.LLMProvider.OPENAI.value:
+            test_openai_llm_connection(
+                api_key=llm_config["apiKey"],
+                model=llm_config["model"],
+            )
+        elif llm_config["llmIdentifier"] == enums.LLMProvider.AZURE.value:
+            test_azure_llm_connection(
+                api_key=llm_config["apiKey"],
+                model=llm_config["model"],
+                base_endpoint=llm_config["apiBase"],
+                api_version=llm_config["apiVersion"],
+            )
+        else:
+            raise LlmResponseError(
+                "LLM Identifier must be either Open AI or Azure, got: "
+                + llm_config["llmIdentifier"]
+            )
+    except AssertionError:
+        raise LlmResponseError(
+            "API version format must be YYYY-MM-DD, got: " + llm_config["apiVersion"]
+        )
+    except requests.exceptions.RequestException:
+        raise LlmResponseError(
+            "Encountered Exception when trying LLM connection: "
+            + traceback.format_exception(*sys.exc_info())[-1]
+        )
+
+
+def prepare_llm_response_code(
+    attribute_item: Attribute,
+    llm_playground_config: Union[Dict[str, Any], None] = None,
+    llm_ac_cache_access_link: Union[str, None] = None,
+    llm_ac_cache_file_upload_link: Union[str, None] = None,
+    num_workers: int = 100,
+    max_api_call_retries: int = 5,
+    retry_sleep_seconds: int = 5,
+) -> str:
+    global LLM_RESPONSE_TMPL_PATH
+    with open(LLM_RESPONSE_TMPL_PATH, "r") as file:
+        lines = [line.rstrip() for line in file if line[0] != "#"]
+
+    llm_code = "\n".join(lines)
+    source_code = attribute_item.source_code
+
+    # llm_playground_config is only set if `run-llm-playground`` invoked this function
+    if not attribute_item.additional_config and llm_playground_config is None:
+        llm_config = {}
+    elif llm_playground_config is None:
+        llm_config = dict(
+            attribute_item.additional_config.get("llmConfig", {}),
+            llmIdentifier=attribute_item.additional_config["llmIdentifier"],
+            templatePrompt=attribute_item.additional_config["templatePrompt"],
+            questionPrompt=attribute_item.additional_config["questionPrompt"],
+            llmAcCacheAccessLink=llm_ac_cache_access_link,
+            llmAcCacheFileUploadLink=llm_ac_cache_file_upload_link,
+        )
```

Comment:
Suggestion for simpler condition

```suggestion
    if llm_playground_config is None:
        if not attribute_item.additional_config:
            llm_config = {}
        else
            llm_config = dict(
                attribute_item.additional_config.get("llmConfig", {}),
                llmIdentifier=attribute_item.additional_config["llmIdentifier"],
                templatePrompt=attribute_item.additional_config["templatePrompt"],
                questionPrompt=attribute_item.additional_config["questionPrompt"],
                llmAcCacheAccessLink=llm_ac_cache_access_link,
                llmAcCacheFileUploadLink=llm_ac_cache_file_upload_link,
            )
```